// lamb-cli/build/config.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
class Config {
    constructor() {
        this.configFilename = "lamb-cli.config.json";
    }
    static get instance() {
        if (!this._instance)
            this._instance = new Config();
        return this._instance;
    }
    get packageJson() {
        return (() => __awaiter(this, void 0, void 0, function* () {
            return JSON.parse(yield promises_1.default.readFile(path_1.default.resolve(__dirname, "..", "package.json"), {
                encoding: "utf-8",
            }));
        }))();
    }
    get cliName() {
        return (() => __awaiter(this, void 0, void 0, function* () {
            return (yield this.packageJson).name;
        }))();
    }
    get cliVersion() {
        return (() => __awaiter(this, void 0, void 0, function* () {
            return (yield this.packageJson).version;
        }))();
    }
    get description() {
        return (() => __awaiter(this, void 0, void 0, function* () {
            return (yield this.packageJson).description;
        }))();
    }
    get workingPath() {
        return process.cwd();
    }
    get metadataPath() {
        return path_1.default.resolve(this.workingPath, this.configFilename);
    }
    get metadataContent() {
        return JSON.stringify({
            functions: [],
        });
    }
    get metadata() {
        return (() => __awaiter(this, void 0, void 0, function* () {
            try {
                yield promises_1.default.access(this.metadataPath);
            }
            catch (error) {
                yield promises_1.default.writeFile(this.metadataPath, this.metadataContent);
            }
            return JSON.parse(yield promises_1.default.readFile(this.metadataPath, { encoding: "utf-8" }));
        }))();
    }
    updateMetadata(updater) {
        return __awaiter(this, void 0, void 0, function* () {
            return promises_1.default.writeFile(this.metadataPath, JSON.stringify(updater(yield this.metadata)));
        });
    }
}
exports.config = Config.instance;
