// lamb-cli/build/lambda-local.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaLocal = void 0;
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const bundler_1 = __importDefault(require("./lib/bundler"));
const zip_1 = __importDefault(require("./lib/zip"));
const tsBP = `export const handler = async ()=> {
  console.log("hello from handler");
}`;
const jsBP = `exports.handler = async()=> {
  console.log("hello from handler");
} 
`;
class LambdaLocal {
    constructor(config) {
        this.config = config;
    }
    createEntryFile({ dir = "", filename, typescript, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const ext = typescript ? "ts" : "js";
            const bp = typescript ? tsBP : jsBP;
            const filePath = path_1.default.resolve(this.config.projectRoot, dir, `${filename}.${ext}`);
            yield promises_1.default.writeFile(filePath, bp);
            return filePath;
        });
    }
    getZipUnit8Array({ entryFile, handlerFileName = "index.js", }) {
        return __awaiter(this, void 0, void 0, function* () {
            const bundler = new bundler_1.default();
            const zipMaker = new zip_1.default();
            const buffer = yield bundler.bundleToBuffer({
                entryFile,
            });
            const zipBuff = zipMaker.createFromBuffer({
                buffer,
                fileName: handlerFileName,
            });
            return Uint8Array.from(zipBuff);
        });
    }
}
exports.LambdaLocal = LambdaLocal;
