// lamb-cli/build/lambda.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteExecution = exports.CreateExecution = void 0;
const client_lambda_1 = require("@aws-sdk/client-lambda");
const uuid_1 = require("uuid");
const role_1 = require("./role");
function timeOut(milSec) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((res) => {
            setTimeout(res, milSec);
        });
    });
}
class LambdaExecution {
    constructor(config) {
        this.config = config;
        this.client = new client_lambda_1.LambdaClient(config);
    }
}
class CreateExecution extends LambdaExecution {
    execute({ name, unit8ArrayBuffer, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const { arn: RoleArn, name: RoleName } = yield (0, role_1.createLambdaExecutionRole)({
                RoleName: `${name}-role-${(0, uuid_1.v4)()}`,
                region: this.config.region,
                credentials: this.config.credentials,
            });
            yield timeOut(10000);
            const command = new client_lambda_1.CreateFunctionCommand({
                FunctionName: name,
                Role: RoleArn,
                Runtime: client_lambda_1.Runtime.nodejs16x,
                Handler: `index.handler`,
                Code: {
                    ZipFile: unit8ArrayBuffer,
                },
            });
            try {
                const { FunctionArn } = yield this.client.send(command);
                return {
                    FunctionArn,
                    RoleArn,
                    RoleName,
                };
            }
            catch (error) {
                if (RoleName) {
                    yield (0, role_1.deleteRole)({
                        RoleName,
                        region: this.config.region,
                        credentials: this.config.credentials,
                    });
                }
                throw error;
            }
        });
    }
}
exports.CreateExecution = CreateExecution;
class DeleteExecution extends LambdaExecution {
    execute({ FunctionArn, RoleName, }) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.send(new client_lambda_1.DeleteFunctionCommand({
                FunctionName: FunctionArn,
            }));
            yield (0, role_1.deleteRole)({
                RoleName,
            });
        });
    }
}
exports.DeleteExecution = DeleteExecution;
