// lamb-cli/build/metadata.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Metadata = exports.functions = void 0;
const promises_1 = require("fs/promises");
const config_1 = require("./config");
const add = (fnMt) => __awaiter(void 0, void 0, void 0, function* () {
    yield config_1.config.updateMetadata((m) => {
        m.functions.push(fnMt);
        return m;
    });
});
const get = (name) => __awaiter(void 0, void 0, void 0, function* () {
    const mt = yield config_1.config.metadata;
    return mt.functions.find((f) => f.name === name);
});
exports.functions = {
    add,
    get,
};
class Metadata {
    constructor(config) {
        this.config = config;
    }
    get fileContent() {
        return (() => __awaiter(this, void 0, void 0, function* () {
            return JSON.parse(yield (0, promises_1.readFile)(this.config.filePath, { encoding: "utf-8" }));
        }))();
    }
    saveFileContent(arg) {
        return __awaiter(this, void 0, void 0, function* () {
            let newMt;
            if (typeof arg === "function") {
                const mt = yield this.fileContent;
                newMt = arg(mt);
            }
            else
                newMt = arg;
            yield (0, promises_1.writeFile)(this.config.filePath, JSON.stringify(newMt));
        });
    }
    addFunction(fnMt) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.saveFileContent((mt) => {
                mt.functions.push(fnMt);
                return mt;
            });
        });
    }
    removeFunction(name) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.saveFileContent((mt) => {
                mt.functions = mt.functions.filter((f) => f.name !== name);
                return mt;
            });
        });
    }
    getFunctions() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.fileContent).functions;
        });
    }
    static init({ filePath }) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield (0, promises_1.access)(filePath);
            }
            catch (error) {
                yield (0, promises_1.writeFile)(filePath, JSON.stringify({
                    functions: [],
                }));
            }
            return new Metadata({ filePath });
        });
    }
}
exports.Metadata = Metadata;
