// lamb-cli/build/__test__/lambda-local.test.js
"use strict";
/**
 * Features
 *
 * create a file in the specified folder
 * directory should be there
 *
 * Bundle the code from entry file
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const memfs_1 = require("memfs");
const lambda_local_1 = require("../lambda-local");
const bundler_1 = __importDefault(require("../lib/bundler"));
const zip_1 = __importDefault(require("../lib/zip"));
jest.mock("../lib/zip.ts");
jest.mock("../lib/bundler.ts");
jest.mock("fs/promises", () => memfs_1.fs.promises);
describe("Test for lambda local artifacts", () => {
    beforeAll(() => {
        memfs_1.vol.reset();
        bundler_1.default.mockClear();
        zip_1.default.mockClear();
    });
    test("Should create an entry file", () => __awaiter(void 0, void 0, void 0, function* () {
        const c = new lambda_local_1.LambdaLocal({ projectRoot: "/" });
        yield c.createEntryFile({
            filename: "test",
            typescript: true,
        });
        yield c.createEntryFile({
            filename: "test",
            typescript: false,
        });
        expect(memfs_1.vol.toJSON()["/test.ts"]).toBeDefined();
        expect(memfs_1.vol.toJSON()["/test.js"]).toBeDefined();
    }));
    test("Should get the unit8arraybuffer of the zip", () => __awaiter(void 0, void 0, void 0, function* () {
        const bundleToBuffer = jest.spyOn(bundler_1.default.prototype, "bundleToBuffer");
        const createFromBufferMock = jest
            .spyOn(zip_1.default.prototype, "createFromBuffer")
            .mockReturnValue(Buffer.from("Hello"));
        const lambdaLocal = new lambda_local_1.LambdaLocal({
            projectRoot: "/",
        });
        const resUnit8Array = yield lambdaLocal.getZipUnit8Array({
            entryFile: "/my-lambda.js",
            handlerFileName: "index.js",
        });
        expect(bundleToBuffer).toHaveBeenCalledWith({
            entryFile: "/my-lambda.js",
            bundlerConfig: undefined,
        });
        expect(createFromBufferMock.mock.calls[0][0].fileName).toBe("index.js");
        expect(Buffer.from(resUnit8Array).toString()).toBe("Hello");
    }));
});
