// lamb-cli/build/__test__/metadata.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const memfs_1 = require("memfs");
const metadata_1 = require("../metadata");
jest.mock("fs/promises", () => memfs_1.fs.promises);
function createTestFnMeta(name = "test-fn") {
    return {
        name,
        arn: "test-arn",
        entryFile: "entry/file.ts",
        region: "ap-south-1",
        roleName: "test-role",
    };
}
describe("Test for metadata-file", () => {
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        memfs_1.vol.reset();
    }));
    test("Should Add the function details properly", () => __awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const metadata = yield metadata_1.Metadata.init({ filePath: "/metadata.json" });
        yield metadata.addFunction(createTestFnMeta("test-fnn"));
        const { functions } = JSON.parse(memfs_1.fs.readFileSync("/metadata.json", { encoding: "utf-8" }));
        expect((_a = functions[0]) === null || _a === void 0 ? void 0 : _a.name).toBe("test-fnn");
    }));
    test("Should Delete the function details properly", () => __awaiter(void 0, void 0, void 0, function* () {
        const metadata = yield metadata_1.Metadata.init({ filePath: "/metadata.json" });
        yield metadata.addFunction(createTestFnMeta("fn-1"));
        yield metadata.addFunction(createTestFnMeta("fn-2"));
        yield metadata.removeFunction("fn-2");
        const { functions } = JSON.parse(memfs_1.fs.readFileSync("/metadata.json", { encoding: "utf-8" }));
        expect(functions.length).toBe(1);
    }));
    test("Should list the functions", () => __awaiter(void 0, void 0, void 0, function* () {
        const metadata = yield metadata_1.Metadata.init({ filePath: "/metadata.json" });
        yield metadata.addFunction(createTestFnMeta("fn-1"));
        yield metadata.addFunction(createTestFnMeta("fn-2"));
        const functions = yield metadata.getFunctions();
        expect(functions.length).toBe(2);
    }));
});
